### ISSUE TITLE
re-sampling to specified interval (second, minute, etc.)

### ISSUE THREAD

**Issue by erigler-usgs** [usgs/geomag-algorithms#235]
If geomag-algorithms wants to correct for bad time stamps that may have made it into Edge, this seems like a good place to start:

https://github.com/usgs/geomag-algorithms/blob/5ead37604866a470235e0c3382e6ab2d3971fa7b/geomagio/edge/EdgeFactory.py#L475-L477

This call to get_waveforms() should be modified to add `cleanup=False` in order to prevent the default ObsPy cleanup behavior, which can lead to incorrect time stamps . The returned stream will now contain multiple traces with identical SCNLs that must be merged "correctly".

With access to `interval`, `starttime` and `endtime`, we can determine the "correct" time stamps (top of second, minute, etc.), then use Stream.interpolate() to re-sample each trace appropriately.

Finally, Stream.merge() should be used to combine the traces with identical SCNLs.

This issue is mostly to serve as a bookmark for myself if/when I can make time to implement and test everything, but if anyone has any comments or alternative suggestions, please share.

**Comment by jmfee-usgs** [comment]
I think we shouldn't automatically fix the timestamps, but switching the cleanup flag so we can better detect any issues that exist seems like a good idea.

Also, we're using a local version of the waveserver client and we may want to switch back to the obspy library version to get future updates.  We patched a local copy, but that patch has now been merged into upstream.

**Comment by erigler-usgs** [comment]
Perhaps an option to "fix the timestamps"?

The long and short of it is that any algorithm that deals with more than a single data point at a time should ensure that time stamps fall when they are expected. This includes SqDist and filter to start, but I'm sure there will be more.

A specific, but by no means isolated, example is a problem that cropped up with recent changes to SqDistAlgorithm.process() (actually the problem was always there, but now process() correctly throws an exception). The comparison of the algorithm object's `next_starttime` and the input trace's `starttime` fails. The two don't match because the long 90-day trace currently (as of Nov. 13, 2018) returned to initialize the algorithm for the SHU observatory has faulty time stamps ending on 28 seconds, when they should fall at the top of the minute.

TLDR: this problem arises because there is a short interval on August 12 where some of the time stamps actually do fall on the 28 second mark, or at least they do for XY data stored on Edge. This interval begins at 2018-08-12T11:27:28.000, and ends at 2018-08-12T12:19:28.000. If you pull a short interval that is entirely before that, time stamps are good. If you pull one that is entirely after that, time stamps are good. If you pull one that starts within that interval, time stamps are bad (28 seconds).

 Further, at least in this particular case, times just prior to this interval actually include a gap, which means the first traceBuf returned by Edge with actual data includes the bad starttime, which corrupts the remainder of the trace because of the way get_waveforms works by default, using the first startime to define the entire trace, even when it was constructed from multiple traceBufs.

Oddly, this problem seems isolated to XY data, and does not seem to be a problem with HE data. Or maybe this is a hint at a more fundamental problem, I'm not sure.

**Comment by erigler-usgs** [comment]
Just a small addendum to my previous TLDR:

It seems that the PHP webservice (https://geomag.usgs.gov/ws/edge/) gets this at least partially correct in that it reports the 28 second time stamps only for the interval that actually has 28 second time stamps stored on Edge.

If it wasn't clear previously, geomag-algorithms will apply the 28 second offset to all samples in the trace, even if the sample came from a perfectly good (i.e., no 28 second offset) tracebuff.

**Comment by erigler-usgs** [comment]
@jmfee-usgs...just kicking this issue so you can find it easily. For now, I'd just like you to look at the interval noted above on mage2, and maybe we can figure out what went wrong at the (Edge) source before trying to implement a fix in geomag-algorithms. Thx.
